src/term/transform/scott_generation.rs

```diff
@@ -0,0 +1,84 @@
+use crate::term::{Book, Definition, Name, Rule, Term};
+
+impl Book {
+  pub fn generate_scott_adts(&mut self) {
+    for adt in self.adts.values() {
+      let curr_ctrs = adt.ctrs.clone();
+
+      let mut ctr_args = Vec::new();
+      let mut ctrs = Vec::new();
+
+      for (ctr_name, arity) in &curr_ctrs {
+        for ctr in curr_ctrs.keys() {
+          ctrs.push(ctr.clone());
+          if ctr == ctr_name {
+            make_args(arity, &mut ctr_args);
+          }
+        }
+
+        let lam = make_lam(&ctr_args, &mut ctrs, ctr_name);
+
+        ctr_args.clear();
+        ctrs.clear();
+
+        let def_id = self.def_names.insert(ctr_name.clone());
+
+        let rule = Rule { def_id, pats: vec![], body: lam };
+        let rules = vec![rule];
+        let def = Definition { def_id, rules };
+
+        self.defs.insert(def_id, def);
+      }
+    }
+  }
+}
+
+fn make_args(args_count: &usize, ctr_args: &mut Vec<Name>) {
+  for n in 0 .. *args_count {
+    let nam = Name(format!("a{n}"));
```

Comment:
Maybe we could use the names given by the user in the program.
Would look a bit better, but we gotta change the ADT.

Reply:
Actually, if the user pulls a funny move and declares something like this `data = (Ctr Ctr)` it would break, so we can leave it like this